examples/fluids/navierstokes.c

```diff
@@ -1200,64 +1204,90 @@ int main(int argc, char **argv) {
     CeedQFunctionAddInput(qf_ifunctionSur, "dq", ncompq*dim, CEED_EVAL_GRAD); //assumed volumetric elements
     CeedQFunctionAddInput(qf_ifunctionSur, "qdataSur", qdatasizeSur, CEED_EVAL_NONE);
     CeedQFunctionAddInput(qf_ifunctionSur, "x", ncompx, CEED_EVAL_INTERP);
-    CeedQFunctionAddInput(qf_rhsSur, "qdata", qdatasizeVol, CEED_EVAL_NONE);
+    CeedQFunctionAddInput(qf_ifunctionSur, "qdata", qdatasizeVol, CEED_EVAL_NONE); //assumed volumetric elements
     CeedQFunctionAddOutput(qf_ifunctionSur, "v", ncompq, CEED_EVAL_INTERP);
   }
-
-  // Create the operator that builds the quadrature data for the NS operator
-  CeedOperatorCreate(ceed, qf_setupSur, NULL, NULL, &op_setupSur);
-  CeedOperatorSetField(op_setupSur, "dx", restrictxSur, basisxSur, CEED_VECTOR_ACTIVE);
-  CeedOperatorSetField(op_setupSur, "weight", CEED_ELEMRESTRICTION_NONE,
-                       basisxSur, CEED_VECTOR_NONE);
-  CeedOperatorSetField(op_setupSur, "qdataSur", restrictqdiSur,
+  // Create CEED Operator for each face
+  for(CeedInt i=0; i<numOutFlow; i++){
+
+    ierr = GetRestrictionForDomain(ceed, dm, ncompx, dimSur, height, domainLabel, bc.outflow[i], numP_Sur,
+                                   numQ_Sur, qdatasizeSur, &restrictqSur[i], &restrictxSur[i],
+                                   &restrictqdiSur[i]); CHKERRQ(ierr);
+    // Create the CEED vectors that will be needed in setup
+    CeedElemRestrictionGetNumElements(restrictqSur[i], &localNelemSur[i]);
+    CeedVectorCreate(ceed, qdatasizeSur*localNelemSur[i]*NqptsSur, &qdataSur[i]);
+    CeedVectorCreate(ceed, qdatasizeSur*localNelemSur[i]*NqptsSur, &qdataSur_[i]);
+
+    // Create the operator that builds the quadrature data for the NS operator
+    CeedOperatorCreate(ceed, qf_setupSur, NULL, NULL, &op_setupSur[i]);
+    CeedOperatorSetField(op_setupSur[i], "dx", restrictxSur[i], basisxSur, CEED_VECTOR_ACTIVE);
+    CeedOperatorSetField(op_setupSur[i], "weight", CEED_ELEMRESTRICTION_NONE,
+                         basisxSur, CEED_VECTOR_NONE);
+    CeedOperatorSetField(op_setupSur[i], "qdataSur", restrictqdiSur[i],
+                         CEED_BASIS_COLLOCATED, CEED_VECTOR_ACTIVE);
+
+    // Utility operator that builds the quadrature data for computing viscous terms
+    CeedOperatorCreate(ceed, qf_setupVol, NULL, NULL, &op_setupSur_[i]);
+    CeedOperatorSetField(op_setupSur_[i], "dx", restrictxSur[i], basisxVol, CEED_VECTOR_ACTIVE);
+    CeedOperatorSetField(op_setupSur_[i], "weight", CEED_ELEMRESTRICTION_NONE,
+                         basisxVol, CEED_VECTOR_NONE);
+    CeedOperatorSetField(op_setupSur_[i], "qdata", restrictqdiSur[i],
```

Comment:
@jedbrown , I remember you were suggesting assuming the surface elements as 3d (in our 3d examples) and dropping the quadrature data on the surface. I was thinking of creating a separate restriction for the faces with `height = 0` for this block of code but the compiler complained (is it even a feasible think to do with `CreateRestrictionFromPlex()`?) so I decided to use the surface restrictions (which I don't think it's correct!).

Reply:
Is there a Plex limitation?

Reply:
@knepley , do you mean Plex limitation for creating 3d restrictions on the faces? I have to build the 3d restriction again to answer positively (right now I am having trouble building PETSc) but if I remember correctly, the limitation was in libCEED and the compatibility with the operators.